### DISCUSSION TITLE
Valid query containing `UNIQUE NULLS NOT DISTINCT` is not validated

### DISCUSSION THREAD

**Discussion by data-envoy** [supabase/supabase#16991] [Contribute to Supabase]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

In Postgres 15, the `NULLS NOT DISTINCT` syntax was introduced. However, queries including this syntax fail in Studio - both on CLI and app.supabase.com

## To Reproduce

1. Create a new database, either via app.supabase.com or with supabase CLI
2. Go to the SQL editor
3. Run a command that includes the `NULLS NOT DISTINCT` syntax
    ```
    CREATE TABLE new_style (
      id BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
      val1 TEXT NOT NULL,
      val2 TEXT NULL,
      CONSTRAINT uq_val1_val2_new UNIQUE NULLS NOT DISTINCT (val1, val2)
    );
   ```    
   It will fail with `Failed to validate sql query: syntax error at or near "NULLS"`
4. Or create a table and try to add a constraint
    ```
    CREATE TABLE new_style (
      id BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
      val1 TEXT NOT NULL,
      val2 TEXT NULL
    );
    ALTER TABLE new_style ADD CONSTRAINT uq_val1_val2_new UNIQUE NULLS NOT DISTINCT (val1, val2);
    ```
5. Connect to the database with a client like Postico, and the command will work ok.

## Expected behavior

It is expected that the SQL Editor validates the `NULLS NOT DISTINCT` syntax correctly.

## Additional context

There are 2 reports in discord but no bug reports here.

- https://discord.com/channels/839993398554656828/1056281310458544239/1056281310458544239
- https://discord.com/channels/839993398554656828/1056281310458544239/1056285615471214643

**Comment by nith2001** [comment]
**DISCLAIMER:** New to Supabase open-source, hoping to contribute sooner or later!

Started to take a look at how Postgres database queries get validated in Supabase. After some digging, I narrowed it down to `supabase/studio/stores/pgmeta/MetaStore.ts` on Line 243 where it has the following:

```code
async validateQuery(value: string) {
    try {
      const headers: any = { 'Content-Type': 'application/json' }
      if (this.connectionString) headers['x-connection-encrypted'] = this.connectionString
      const url = `${this.baseUrl}/query/validate`
      const response = await post(url, { query: value }, { headers })
      if (response.error) throw response.error

      return response
    } catch (error: any) {
      return { error }
    }
}
```

The URL being used here seems to be structured as `${API_URL}/pg-meta/${projectRef}/query/validate`. To me, the `pg-meta` reference here seemed to indicate something related to the `supabase/postgres-meta` repository, so I started looking in that direction but couldn't find any reference to a `validate` endpoint. Am I looking in the wrong direction? Any help is appreciated!

**Comment by data-envoy** [comment]
It looks like the right direction. But, likewize, I'm new to Supabase. Are you running the supabase cli? I wonder is there a debug or verbose flag or env var that can be set on `supabase start`? That may give a stack trace for the validation error, either visible in the SQL Editor UI or maybe the browser Dev tools > Network inspector.

**Comment by data-envoy** [comment]
You're right, it's the postgres-meta repo.

Here's a trace of the error.

```
{
  error: Error: syntax error at or near "NULLS"
      at parseQuerySync (/usr/src/app/node_modules/libpg-query/index.js:21:31)
      at parse (/usr/src/app/node_modules/pgsql-parser/main/index.js:52:47)
      at Module.Parse (file:///usr/src/app/dist/lib/Parser.js:19:22)
      at Object.<anonymous> (file:///usr/src/app/dist/server/routes/query.js:36:40)
      at preHandlerCallback (/usr/src/app/node_modules/fastify/lib/handleRequest.js:128:37)
      at preValidationCallback (/usr/src/app/node_modules/fastify/lib/handleRequest.js:112:5)
      at handler (/usr/src/app/node_modules/fastify/lib/handleRequest.js:76:7)
      at /usr/src/app/node_modules/fastify/lib/contentTypeParser.js:192:9
      at AsyncResource.runInAsyncScope (node:async_hooks:203:9)
      at done (/usr/src/app/node_modules/fastify/lib/contentTypeParser.js:186:14) {
    fileName: 'scan.l',
    functionName: 'scanner_yyerror',
    lineNumber: 1236,
    cursorPosition: 161,
    context: null
  }

```

The function is [here](https://github.com/supabase/postgres-meta/blob/230fe4020cf39e5ec02c64cb7dded8011950a027/src/lib/Parser.ts#L19).

Dependency of pgsql-parser [is here](https://github.com/supabase/postgres-meta/blob/230fe4020cf39e5ec02c64cb7dded8011950a027/package.json#L62).

In turn pgsql-parser has a dep of `"libpg-query": "13.3.1"` [here](https://github.com/pyramation/pgsql-parser/blob/3d78dc66d800eb8cdf76851337f27b972160fb41/packages/parser/package.json#L75).

The newest version of libpg-query is [15.02](https://www.npmjs.com/package/libpg-query).

So I *think* that is the cause of this issue. Outdated sub dependency.

Edit, the [pgsql-parser readme](https://github.com/pyramation/pgsql-parser) also confirms the compatibility up to postgres 13.

**Comment by nith2001** [comment]
Woooo, that's great! I actually don't have my dev environment fully set up yet so I've been playing it by ear (or I guess eye) to trace the issue down. 

I'll see if I can work on this.

**Comment by data-envoy** [comment]
No problem, it was certainly helpful to have your direction to follow. 

I've mentioned this issue https://github.com/pyramation/pgsql-parser/issues/93 . The issue was created by the project owner, so I assume it's on his todo list. 

No idea how much work it is for him, maybe a version bump in package.json and run tests, maybe more involved, idk.

**Comment by nith2001** [comment]
@data-envoy By the way, how did get the stack trace? I am not currently running the Supabase CLI and I'm not sure how it helps here. To start the codebase, I just do a `docker compose up` to start the site. Was it some type of debug flag or something you did?

Edit: Essentially I'm asking because I don't know how you managed the rest of the debug process, especially because I don't quite have `pg-meta` repo setup and idk how to get `pg-meta` and `supabase` repos linked up on my dev environment.

**Comment by data-envoy** [comment]
I didn't really do it the proper way by cloning and running postgres-meta repo. So please don't mind the hacky-ness.

- With the supabase cli I ran `supabase start --debug`.
- That starts up a a bunch of containers, I think it's similar to docker-compose up with a bunch of config behind the scenes.
- View logs of the `postgres-meta` container (I cheat by using Docker for VSCode GUI)
- In SQL Editor run the problem query
- Log out is: `{"level":"error","time":"2023-04-03T08:29:51.756Z","pid":19,"hostname":"52cb37d69a68","reqId":"req-2m","error":{"fileName":"scan.l","functionName":"scanner_yyerror","lineNumber":1236,"cursorPosition":161,"context":null},"request":{"method":"POST","url":"/query/parse","pg":"supabase_db_xyz","opt":""}}`
- Inspect the `postgres-meta` container (VSCode GUI again) to find `"WorkingDir": "/usr/src/app",`
- Attach VSCode to the `postgres-meta` container and open dir `/usr/src/app`.
- Search for files named parse and find Parser.js.
- Add my own console.log to each of 3 catch statements. 
- Restart the container, because I don't exactly know how to stop the running nodejs process.
- In SQL Editor re-run the problem query. And the trace is logged.

**Comment by nith2001** [comment]
Wow, that makes sense. For some reason, it didn't click for me that running the Supabase codebase involved deploying the `postgres-meta` as a container that was usable, but now that I think about it, it's pretty obvious because that's the only way that all of the database support is functional. Thank you!

**Comment by data-envoy** [comment]
No problem... happy I could help shed some light on it.

**Comment by nith2001** [comment]
Found a page that details all of the changes since Postgres 13 to 15, and there might be some potentials issues that will pop up as a result of trying to update to 15 in case just doing a library update is not enough: https://www.postgresql.org/about/featurematrix/

Edit: Trying to contribute to the `pgsql-parser` repo right now but facing some [issues](https://github.com/pyramation/pgsql-parser/issues/93#issuecomment-1495139311) with actually understanding how dependency version control is managed in that repo, we'll just have to wait and see if any support can be given by the author.

Edit 2: Resolved the weird issues, now just trying to see how to test if the parser can handle parsing the new feature queries.

**Comment by nith2001** [comment]
Until this gets fixed, it might be a good idea to put somewhere in our documentation that Supabase only provides stable support of Postgres up to Version 13.

**Comment by nith2001** [comment]
Welp, it looks like `libpg-query` is not able to handle `UNIQUE NULLS NOT DISTINCT` yet either. I tried to parse the following table within `pgsql-parser`:

```
CREATE TABLE new_style (
  id BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  val1 TEXT NOT NULL,
  val2 TEXT NULL,
  CONSTRAINT uq_val1_val2_new UNIQUE NULLS NOT DISTINCT (val1, val2)
);
```

And I got the following error:
```
syntax error at or near "NULLS"

      19 | export const parse = (sql) => {
      20 |   if (!sql) throw new Error('no SQL provided to parser');
    > 21 |   const result = parseQuerySync(sql);
         |                                ^
      22 |   return result.stmts.map(mapStmt);
      23 | };
      24 |

      at parseQuerySync (../../node_modules/libpg-query/index.js:21:31)
      at parse (src/index.js:21:32)
      at check (__tests__/kitchen-sink.test.js:13:21)
      at Object.<anonymous> (__tests__/kitchen-sink.test.js:42:5)
```
I'll throw it in their issue board, but according to their repo, it seems parsing Postgres 15 is still in active development, but I'll try to verify that just in case. But if this is just some feature that `libpg-query` needs to develop before anything can work, then I'm afraid this is outside of my expertise until that's done.

**Comment by data-envoy** [comment]
Thanks for the investigation on this @nith2001 .

Maybe this is a stupid question... do queries from the SQL Editor **always** need to be parsed & validated?

Why not let the query hit the database, catch and return an error from there?

**Comment by data-envoy** [comment]
@nith2001 If you check the [script/buildAddon.sh](https://github.com/pyramation/libpg-query-node/pull/27/files#diff-445661ab042c30bc23d38d8656deaf29c52413ac9b108958ced3ff1175bcd1e1) on the most recent release of `libpg-query`, it has a commit hash for the dependency on [pganalyze/libpg_query](https://github.com/pganalyze/libpg_query).

That commit was made in Dec and [one release](https://github.com/pganalyze/libpg_query/commits/15-latest) has been made since then.

I'm short on time, so sorry to leave at this, but if you were so inclined... you could try `libpg-query` locally with an updated commit hash in `script/buildAddon.sh`.

**Comment by nith2001** [comment]
@data-envoy I guessing it's because we don't want to overrun the database with non-functional queries, thereby marginally reducing the load on the database? I can't think of a better reason.

**Comment by data-envoy** [comment]
Sure, that makes sense.
